File: backend/solar_calculations.py
import math
import numpy as np
import logging
from typing import Dict, List, Optional
from config.constants import PANEL_SPECS, SYSTEM_LOSSES, FINANCIAL_PARAMS

class SolarCalculator:
    """Comprehensive solar potential and financial calculations"""
    
    def __init__(self):
        self.logger = logging.getLogger(__name__)
    
    def calculate_potential(self, roof_metrics: Dict, solar_data: Dict, 
                          panel_type: str, electricity_rate: float, 
                          installation_cost_per_watt: float) -> Dict:
        """
        Calculate complete solar potential including energy and financial analysis
        
        Args:
            roof_metrics: Roof analysis results
            solar_data: Solar irradiance data from NASA
            panel_type: Selected panel type
            electricity_rate: Local electricity rate ($/kWh)
            installation_cost_per_watt: Installation cost per watt
        
        Returns:
            Dictionary containing solar potential calculations
        """
        
        try:
            # Get panel specifications
            panel_specs = PANEL_SPECS.get(panel_type, PANEL_SPECS['monocrystalline'])
            
            # Calculate system size
            system_size_kw = self._calculate_system_size(roof_metrics, panel_specs)
            
            # Calculate energy production
            annual_energy_kwh = self._calculate_annual_energy(
                system_size_kw, solar_data, roof_metrics, panel_specs
            )
            
            # Calculate financial metrics
            financial_metrics = self._calculate_financial_metrics(
                system_size_kw, annual_energy_kwh, electricity_rate, 
                installation_cost_per_watt
            )
            
            # Calculate environmental impact
            environmental_impact = self._calculate_environmental_impact(annual_energy_kwh)
            
            return {
                'system_size_kw': system_size_kw,
                'panel_count': int(system_size_kw * 1000 / panel_specs['power_watts']),
                'annual_energy_kwh': annual_energy_kwh,
                'monthly_energy_kwh': annual_energy_kwh / 12,
                'daily_energy_kwh': annual_energy_kwh / 365,
                'capacity_factor': self._calculate_capacity_factor(
                    annual_energy_kwh, system_size_kw
                ),
                **financial_metrics,
                **environmental_impact,
                'system_specifications': {
                    'panel_type': panel_type,
                    'panel_power': panel_specs['power_watts'],
                    'panel_efficiency': panel_specs['efficiency'],
                    'system_losses': SYSTEM_LOSSES
                }
            }
            
        except Exception as e:
            self.logger.error(f"Solar potential calculation failed: {str(e)}")
            return self._get_default_potential()
    
    def _calculate_system_size(self, roof_metrics: Dict, panel_specs: Dict) -> float:
        """Calculate optimal system size in kW"""
        
        try:
            usable_area = roof_metrics['usable_area']  # m²
            panel_area = panel_specs['area_m2']
            panel_power = panel_specs['power_watts']
            
            # Calculate number of panels that can fit
            max_panels = int(usable_area / panel_area)
            
            # Apply spacing factor (typically 70-80% of theoretical maximum)
            spacing_factor = 0.75
            actual_panels = int(max_panels * spacing_factor)
            
            # Calculate system size in kW
            system_size_kw = (actual_panels * panel_power) / 1000
            
            # Cap at reasonable residential system size (50kW max)
            system_size_kw = min(system_size_kw, 50.0)
            
            return max(1.0, system_size_kw)  # Minimum 1kW system
            
        except Exception as e:
            self.logger.error(f"System size calculation failed: {str(e)}")
            return 5.0  # Default 5kW system
    
    def _calculate_annual_energy(self, system_size_kw: float, solar_data: Dict, 
                               roof_metrics: Dict, panel_specs: Dict) -> float:
        """Calculate annual energy production in kWh"""
        
        try:
            # Get average daily solar irradiance (kWh/m²/day)
            if 'annual_irradiance' in solar_data:
                daily_irradiance = solar_data['annual_irradiance'] / 365
            elif 'monthly_irradiance' in solar_data:
                daily_irradiance = np.mean(solar_data['monthly_irradiance'])
            else:
                # Default irradiance for moderate climate
                daily_irradiance = 4.5  # kWh/m²/day
            
            # Apply orientation factor
            orientation_factor = self._get_orientation_factor(roof_metrics['orientation'])
            
            # Apply tilt factor
            tilt_factor = self._get_tilt_factor(roof_metrics['slope'])
            
            # Apply shading factor
            shading_factor = 1.0 - roof_metrics['shading_factor']
            
            # System performance ratio (accounts for losses)
            performance_ratio = (
                SYSTEM_LOSSES['inverter_efficiency'] * 
                SYSTEM_LOSSES['dc_losses'] * 
                SYSTEM_LOSSES['ac_losses'] * 
                SYSTEM_LOSSES['soiling_losses']
            )
            
            # Calculate annual energy production
            annual_energy_kwh = (
                system_size_kw * 
                daily_irradiance * 
                orientation_factor * 
                tilt_factor * 
                shading_factor * 
                performance_ratio * 
                365
            )
            
            return max(0, annual_energy_kwh)
            
        except Exception as e:
            self.logger.error(f"Energy calculation failed: {str(e)}")
            return system_size_kw * 1200  # Default 1200 kWh/kW/year
    
    def _get_orientation_factor(self, orientation: str) -> float:
        """Get energy production factor based on roof orientation"""
        
        orientation_factors = {
            'south': 1.00,
            'southeast': 0.95,
            'southwest': 0.95,
            'east': 0.85,
            'west': 0.85,
            'northeast': 0.75,
            'northwest': 0.75,
            'north': 0.60
        }
        
        return orientation_factors.get(orientation.lower(), 0.85)
    
    def _get_tilt_factor(self, slope_degrees: float) -> float:
        """Get energy production factor based on roof tilt"""
        
        # Optimal tilt is typically close to latitude
        # This is a simplified model - optimal tilt varies by location
        if slope_degrees < 10:
            return 0.90  # Too flat
        elif slope_degrees <= 15:
            return 0.95
        elif slope_degrees <= 25:
            return 1.00  # Optimal range
        elif slope_degrees <= 35:
            return 0.98
        elif slope_degrees <= 45:
            return 0.95
        else:
            return 0.85  # Too steep
    
    def _calculate_capacity_factor(self, annual_energy_kwh: float, system_size_kw: float) -> float:
        """Calculate system capacity factor"""
        
        if system_size_kw == 0:
            return 0.0
        
        # Capacity factor = actual energy / theoretical maximum energy
        theoretical_max = system_size_kw * 8760  # kWh (24 hours * 365 days)
        capacity_factor = annual_energy_kwh / theoretical_max
        
        return min(1.0, max(0.0, capacity_factor))
    
    def _calculate_financial_metrics(self, system_size_kw: float, annual_energy_kwh: float,
                                   electricity_rate: float, installation_cost_per_watt: float) -> Dict:
        """Calculate comprehensive financial metrics"""
        
        try:
            # Calculate costs
            total_cost = system_size_kw * 1000 * installation_cost_per_watt
            
            # Apply federal tax incentive (30% as of 2024)
            federal_incentive = total_cost * FINANCIAL_PARAMS['federal_tax_credit']
            net_cost = total_cost - federal_incentive
            
            # Calculate savings
            annual_savings = annual_energy_kwh * electricity_rate
            
            # Account for electricity rate escalation
            escalation_rate = FINANCIAL_PARAMS['electricity_rate_escalation']
            
            # Calculate NPV and payback
            payback_years = self._calculate_payback_period(
                net_cost, annual_savings, escalation_rate
            )
            
            # Calculate 25-year lifetime value
            lifetime_savings = self._calculate_lifetime_savings(
                annual_savings, escalation_rate, 25
            )
            
            # Calculate ROI
            roi_percent = ((lifetime_savings - net_cost) / net_cost) * 100 if net_cost > 0 else 0
            
            # Monthly savings (first year)
            monthly_savings = annual_savings / 12
            
            return {
                'total_cost': total_cost,
                'federal_incentive': federal_incentive,
                'net_cost': net_cost,
                'annual_savings': annual_savings,
                'monthly_savings': monthly_savings,
                'payback_years': payback_years,
                'lifetime_savings': lifetime_savings,
                'roi_percent': roi_percent,
                'cost_per_watt': installation_cost_per_watt,
                'savings_per_kwh': electricity_rate
            }
            
        except Exception as e:
            self.logger.error(f"Financial calculation failed: {str(e)}")
            return self._get_default_financial_metrics()
    
    def _calculate_payback_period(self, initial_cost: float, annual_savings: float, 
                                escalation_rate: float) -> float:
        """Calculate simple payback period with escalation"""
        
        if annual_savings <= 0:
            return 99.0  # No payback
        
        cumulative_savings = 0
        year = 0
        current_annual_savings = annual_savings
        
        while cumulative_savings < initial_cost and year < 50:
            year += 1
            cumulative_savings += current_annual_savings
            current_annual_savings *= (1 + escalation_rate)
        
        return min(year, 50.0)
    
    def _calculate_lifetime_savings(self, annual_savings: float, escalation_rate: float, 
                                  years: int) -> float:
        """Calculate total savings over system lifetime"""
        
        total_savings = 0
        current_annual_savings = annual_savings
        
        for year in range(years):
            total_savings += current_annual_savings
            current_annual_savings *= (1 + escalation_rate)
        
        return total_savings
    
    def _calculate_environmental_impact(self, annual_energy_kwh: float) -> Dict:
        """Calculate environmental benefits of solar system"""
        
        try:
            # CO2 emissions factor (lbs CO2/kWh) - varies by region
            # Using average US grid emissions factor
            co2_factor_lbs_per_kwh = 0.85
            
            # Annual CO2 offset
            annual_co2_offset_lbs = annual_energy_kwh * co2_factor_lbs_per_kwh
            annual_co2_offset_tons = annual_co2_offset_lbs / 2000
            
            # 25-year lifetime offset
            lifetime_co2_offset_tons = annual_co2_offset_tons * 25
            
            # Equivalent trees planted (1 tree absorbs ~48 lbs CO2/year)
            equivalent_trees = annual_co2_offset_lbs / 48
            
            # Equivalent cars off road (average car emits ~4.6 tons CO2/year)
            equivalent_cars = annual_co2_offset_tons / 4.6
            
            return {
                'annual_co2_offset_lbs': annual_co2_offset_lbs,
                'annual_co2_offset_tons': annual_co2_offset_tons,
                'lifetime_co2_offset_tons': lifetime_co2_offset_tons,
                'equivalent_trees_planted': equivalent_trees,
                'equivalent_cars_removed': equivalent_cars
            }
            
        except Exception as e:
            self.logger.error(f"Environmental impact calculation failed: {str(e)}")
            return {
                'annual_co2_offset_lbs': 0,
                'annual_co2_offset_tons': 0,
                'lifetime_co2_offset_tons': 0,
                'equivalent_trees_planted': 0,
                'equivalent_cars_removed': 0
            }
    
    def _get_default_potential(self) -> Dict:
        """Return default solar potential when calculations fail"""
        
        return {
            'system_size_kw': 5.0,
            'panel_count': 16,
            'annual_energy_kwh': 6000,
            'monthly_energy_kwh': 500,
            'daily_energy_kwh': 16.4,
            'capacity_factor': 0.14,
            **self._get_default_financial_metrics(),
            'annual_co2_offset_lbs': 5100,
            'annual_co2_offset_tons': 2.55,
            'lifetime_co2_offset_tons': 63.75,
            'equivalent_trees_planted': 106,
            'equivalent_cars_removed': 0.55,
            'system_specifications': {
                'panel_type': 'monocrystalline',
                'panel_power': 400,
                'panel_efficiency': 0.20,
                'system_losses': SYSTEM_LOSSES
            }
        }
    
    def _get_default_financial_metrics(self) -> Dict:
        """Return default financial metrics when calculations fail"""
        
        return {
            'total_cost': 15000,
            'federal_incentive': 4500,
            'net_cost': 10500,
            'annual_savings': 1200,
            'monthly_savings': 100,
            'payback_years': 8.8,
            'lifetime_savings': 30000,
            'roi_percent': 185.7,
            'cost_per_watt': 3.0,
            'savings_per_kwh': 0.20
        }

File: docs/js/solar-calculations.js
// Solar Calculations Module for Solar AI Platform
class SolarCalculator {
    constructor() {
        this.nasaApiBase = 'https://power.larc.nasa.gov/api/temporal/daily/point';
        this.panelSpecs = {
            monocrystalline: {
                efficiency: 0.22,
                power_per_m2: 220,
                degradation_rate: 0.005,
                cost_multiplier: 1.0
            },
            polycrystalline: {
                efficiency: 0.18,
                power_per_m2: 180,
                degradation_rate: 0.007,
                cost_multiplier: 0.85
            },
            'thin-film': {
                efficiency: 0.12,
                power_per_m2: 120,
                degradation_rate: 0.008,
                cost_multiplier: 0.7
            }
        };
    }

    async calculateSolarPotential(roofAnalysis, config) {
        try {
            // Get solar irradiance data
            const solarData = await this.getSolarIrradiance(config.latitude, config.longitude);
            
            // Calculate system specifications
            const systemSpecs = this.calculateSystemSpecs(roofAnalysis, config);
            
            // Calculate energy production
            const energyProduction = this.calculateEnergyProduction(systemSpecs, solarData, roofAnalysis);
            
            // Calculate financial metrics
            const financialMetrics = this.calculateFinancialMetrics(systemSpecs, energyProduction, config);
            
            // Calculate environmental impact
            const environmentalImpact = this.calculateEnvironmentalImpact(energyProduction);

            return {
                success: true,
                system_size_kw: systemSpecs.systemSizeKw,
                panel_count: systemSpecs.panelCount,
                annual_energy_kwh: energyProduction.annualEnergyKwh,
                monthly_energy_kwh: energyProduction.monthlyEnergyKwh,
                daily_energy_kwh: energyProduction.dailyEnergyKwh,
                total_cost: financialMetrics.totalCost,
                annual_savings: financialMetrics.annualSavings,
                payback_years: financialMetrics.paybackYears,
                lifetime_savings: financialMetrics.lifetimeSavings,
                roi_percent: financialMetrics.roiPercent,
                co2_offset_kg: environmentalImpact.co2OffsetKg,
                trees_equivalent: environmentalImpact.treesEquivalent,
                solar_irradiance: solarData.averageIrradiance,
                capacity_factor: energyProduction.capacityFactor
            };
        } catch (error) {
            console.error('Solar calculation failed:', error);
            return this.getFallbackCalculations(roofAnalysis, config);
        }
    }

    async getSolarIrradiance(latitude, longitude) {
        try {
            // Try to fetch from NASA POWER API
            const currentYear = new Date().getFullYear();
            const startDate = `${currentYear - 1}0101`;
            const endDate = `${currentYear - 1}1231`;
            
            const url = `${this.nasaApiBase}?latitude=${latitude}&longitude=${longitude}&start=${startDate}&end=${endDate}&community=RE&parameters=ALLSKY_SFC_SW_DWN&format=JSON`;
            
            // Use a CORS proxy for NASA API
            const proxyUrl = `https://api.allorigins.win/get?url=${encodeURIComponent(url)}`;
            
            const response = await fetch(proxyUrl);
            const data = await response.json();
            
            if (data.contents) {
                const nasaData = JSON.parse(data.contents);
                const irradianceData = nasaData.properties.parameter.ALLSKY_SFC_SW_DWN;
                
                // Calculate average irradiance
                const values = Object.values(irradianceData);
                const averageIrradiance = values.reduce((sum, val) => sum + val, 0) / values.length;
                
                return {
                    averageIrradiance: averageIrradiance,
                    source: 'NASA POWER API',
                    dataQuality: 'High'
                };
            }
        } catch (error) {
            console.log('NASA API unavailable, using estimated values');
        }
        
        // Fallback: estimate based on location
        return this.estimateIrradiance(latitude, longitude);
    }

    estimateIrradiance(latitude, longitude) {
        // Rough estimates based on latitude
        const absLatitude = Math.abs(latitude);
        let baseIrradiance;
        
        if (absLatitude < 25) {
            baseIrradiance = 6.0; // Tropical regions
        } else if (absLatitude < 35) {
            baseIrradiance = 5.5; // Subtropical regions
        } else if (absLatitude < 45) {
            baseIrradiance = 4.8; // Temperate regions
        } else {
            baseIrradiance = 4.0; // Northern regions
        }
        
        // Adjust for specific regions (simplified)
        if (longitude > -130 && longitude < -60 && latitude > 25 && latitude < 50) {
            // Continental US adjustments
            if (latitude > 35 && longitude > -120 && longitude < -100) {
                baseIrradiance += 0.5; // Southwest US bonus
            }
        }
        
        return {
            averageIrradiance: baseIrradiance,
            source: 'Estimated',
            dataQuality: 'Medium'
        };
    }

    calculateSystemSpecs(roofAnalysis, config) {
        const panelSpec = this.panelSpecs[config.panelType] || this.panelSpecs.monocrystalline;
        const usableArea = roofAnalysis.usable_area;
        
        // Calculate maximum possible system size
        const maxSystemSizeKw = (usableArea * panelSpec.power_per_m2) / 1000;
        
        // Adjust based on system size preference
        let systemSizeKw;
        switch (config.systemSizePreference) {
            case 'maximum':
                systemSizeKw = maxSystemSizeKw * 0.95; // 95% of max
                break;
            case 'budget':
                systemSizeKw = maxSystemSizeKw * 0.6; // 60% of max
                break;
            default: // optimal
                systemSizeKw = maxSystemSizeKw * 0.8; // 80% of max
                break;
        }
        
        // Calculate panel count
        const panelPowerKw = panelSpec.power_per_m2 / 1000;
        const panelCount = Math.floor(systemSizeKw / panelPowerKw);
        
        return {
            systemSizeKw: Math.round(systemSizeKw * 100) / 100,
            panelCount: panelCount,
            panelSpec: panelSpec,
            usableArea: usableArea
        };
    }

    calculateEnergyProduction(systemSpecs, solarData, roofAnalysis) {
        const { systemSizeKw, panelSpec } = systemSpecs;
        const { averageIrradiance } = solarData;
        
        // Calculate capacity factor
        const peakSunHours = averageIrradiance;
        const capacityFactor = (peakSunHours / 24) * panelSpec.efficiency * 100;
        
        // Adjust for roof orientation and slope
        let orientationFactor = 1.0;
        switch (roofAnalysis.orientation.toLowerCase()) {
            case 'south':
                orientationFactor = 1.0;
                break;
            case 'southeast':
            case 'southwest':
                orientationFactor = 0.95;
                break;
            case 'east':
            case 'west':
                orientationFactor = 0.85;
                break;
            case 'north':
                orientationFactor = 0.7;
                break;
            default:
                orientationFactor = 0.9;
        }
        
        // Slope factor (optimal around 30-35 degrees)
        const optimalSlope = 32;
        const slopeDifference = Math.abs(roofAnalysis.slope - optimalSlope);
        const slopeFactor = Math.max(0.8, 1 - (slopeDifference / 100));
        
        // Shading factor
        const shadingFactor = 1 - roofAnalysis.shading_factor;
        
        // Calculate annual energy production
        const annualEnergyKwh = systemSizeKw * peakSunHours * 365 * orientationFactor * slopeFactor * shadingFactor;
        
        return {
            annualEnergyKwh: Math.round(annualEnergyKwh),
            monthlyEnergyKwh: Math.round(annualEnergyKwh / 12),
            dailyEnergyKwh: Math.round(annualEnergyKwh / 365 * 100) / 100,
            capacityFactor: Math.round(capacityFactor * 100) / 100,
            orientationFactor: orientationFactor,
            slopeFactor: slopeFactor,
            shadingFactor: shadingFactor
        };
    }

    calculateFinancialMetrics(systemSpecs, energyProduction, config) {
        const { systemSizeKw, panelSpec } = systemSpecs;
        const { annualEnergyKwh } = energyProduction;
        
        // Calculate total system cost
        const baseCostPerWatt = parseFloat(config.installationCost);
        const adjustedCostPerWatt = baseCostPerWatt * panelSpec.cost_multiplier;
        const totalCost = systemSizeKw * 1000 * adjustedCostPerWatt;
        
        // Calculate annual savings
        const electricityRate = parseFloat(config.electricityRate);
        const annualSavings = annualEnergyKwh * electricityRate;
        
        // Calculate payback period
        const paybackYears = totalCost / annualSavings;
        
        // Calculate 25-year lifetime savings
        const systemLifeYears = 25;
        let lifetimeSavings = 0;
        
        for (let year = 1; year <= systemLifeYears; year++) {
            // Account for panel degradation
            const degradationFactor = Math.pow(1 - panelSpec.degradation_rate, year - 1);
            const yearlyProduction = annualEnergyKwh * degradationFactor;
            
            // Account for electricity rate inflation (assume 3% annually)
            const inflatedRate = electricityRate * Math.pow(1.03, year - 1);
            
            lifetimeSavings += yearlyProduction * inflatedRate;
        }
        
        const netLifetimeSavings = lifetimeSavings - totalCost;
        
        // Calculate ROI
        const roiPercent = (netLifetimeSavings / totalCost) * 100;
        
        return {
            totalCost: Math.round(totalCost),
            annualSavings: Math.round(annualSavings),
            paybackYears: Math.round(paybackYears * 100) / 100,
            lifetimeSavings: Math.round(netLifetimeSavings),
            roiPercent: Math.round(roiPercent * 100) / 100
        };
    }

    calculateEnvironmentalImpact(energyProduction) {
        const { annualEnergyKwh } = energyProduction;
        
        // CO2 offset calculation (average grid emission factor)
        const co2FactorKgPerKwh = 0.4; // kg CO2 per kWh (varies by region)
        const annualCo2OffsetKg = annualEnergyKwh * co2FactorKgPerKwh;
        
        // Trees equivalent (one tree absorbs ~22 kg CO2 per year)
        const treesEquivalent = Math.round(annualCo2OffsetKg / 22);
        
        return {
            co2OffsetKg: Math.round(annualCo2OffsetKg),
            treesEquivalent: treesEquivalent,
            annualCo2OffsetKg: Math.round(annualCo2OffsetKg)
        };
    }

    getFallbackCalculations(roofAnalysis, config) {
        // Provide reasonable estimates when API calls fail
        const systemSizeKw = Math.min(roofAnalysis.usable_area * 0.15, 20); // Conservative estimate
        const annualEnergyKwh = systemSizeKw * 1200; // Conservative production estimate
        const totalCost = systemSizeKw * 1000 * parseFloat(config.installationCost);
        const annualSavings = annualEnergyKwh * parseFloat(config.electricityRate);
        
        return {
            success: true,
            system_size_kw: Math.round(systemSizeKw * 100) / 100,
            panel_count: Math.floor(systemSizeKw * 4), // Assume 250W panels
            annual_energy_kwh: Math.round(annualEnergyKwh),
            monthly_energy_kwh: Math.round(annualEnergyKwh / 12),
            daily_energy_kwh: Math.round(annualEnergyKwh / 365 * 100) / 100,
            total_cost: Math.round(totalCost),
            annual_savings: Math.round(annualSavings),
            payback_years: Math.round((totalCost / annualSavings) * 100) / 100,
            lifetime_savings: Math.round(annualSavings * 20 - totalCost),
            roi_percent: Math.round(((annualSavings * 20 - totalCost) / totalCost) * 100),
            co2_offset_kg: Math.round(annualEnergyKwh * 0.4),
            trees_equivalent: Math.round(annualEnergyKwh * 0.4 / 22),
            solar_irradiance: 4.5,
            capacity_factor: 18.5,
            source: 'Estimated (Offline)'
        };
    }

    // Utility methods
    formatCurrency(amount) {
        return new Intl.NumberFormat('en-US', {
            style: 'currency',
            currency: 'USD',
            minimumFractionDigits: 0,
            maximumFractionDigits: 0
        }).format(amount);
    }

    formatNumber(number) {
        return new Intl.NumberFormat('en-US').format(Math.round(number));
    }

    formatPercentage(percentage) {
        return `${percentage.toFixed(1)}%`;
    }
}

// Export for use in other modules
window.SolarCalculator = SolarCalculator;
